src/main/kotlin/com/team4099/robot2020/subsystems/Feeder.kt

```diff
@@ -0,0 +1,152 @@
+            if (value != field) {
+                inMasterSparkMax.set(value)
+            }
+            field = value
+        }
+    private var feederStopperPower = 0.0
+        set(value) {
+            if (value != field) {
+                stopperTalon.set(ControlMode.PercentOutput, value)
+            }
+            field = value
+        }
+
+    init {
+        inMasterSparkMax.inverted = false
+        inSlaveSparkMax.inverted = true
+
+        stopperTalon.inverted = true
+    }
+
+    enum class FeederState {
+        HOLD, INTAKE, SHOOT, EXHAUST, IDLE
+    }
+
+    override fun outputTelemetry() {
+        SmartDashboard.putString("feeder/feederState", feederState.toString())
+        SmartDashboard.putNumber("feeder/feederInPower", feederInPower)
+        SmartDashboard.putNumber("feeder/feederStopperPower", feederStopperPower)
+    }
+
+    override fun checkSystem() {}
+
+    override fun registerLogging() {
+        HelixLogger.addSource("Feeder Master Motor Power") { inMasterSparkMax.appliedOutput }
+        HelixLogger.addSource("Feeder Slave Motor Power") { inSlaveSparkMax.appliedOutput }
+        HelixLogger.addSource("Feeder Stopper Motor Power") { stopperTalon.motorOutputPercent }
+
+        HelixLogger.addSource("Feeder Master Motor Current") { inMasterSparkMax.outputCurrent }
+        HelixLogger.addSource("Feeder Slave Motor Current") { inSlaveSparkMax.outputCurrent }
+        HelixLogger.addSource("Feeder Stopper Motor current") { stopperTalon.supplyCurrent }
+
+        HelixLogger.addSource("Feeder State") { feederState }
+    }
+
+    override fun zeroSensors() { }
+
+    override fun onStart(timestamp: Double) {
+        feederState = FeederState.HOLD
+    }
+
+    @Synchronized
+    override fun onLoop(timestamp: Double, dT: Double) {
+
+        if (inBeamBroken) {
+            if (inBeamBrokenTimestamp == -1.0) {
+                inBeamBrokenTimestamp = timestamp
+            }
+        } else {
+            if (timestamp - inBeamBrokenTimestamp >= Constants.BeamBreak.IN_BEAM_BROKEN_BALL_TIME) {
+                ballCount++
+            }
+            inBeamBrokenTimestamp = -1.0
+        }
+
+        if (outBeamBroken) {
+            if (outBeamBrokenTimestamp == -1.0) {
+                outBeamBrokenTimestamp = timestamp
+            }
+        } else {
+            if (timestamp - outBeamBrokenTimestamp >= Constants.BeamBreak.OUT_BEAM_BROKEN_BALL_TIME) {
+                ballCount--
+            }
+            outBeamBrokenTimestamp = -1.0
+        }
+
+        when (feederState) {
+            FeederState.INTAKE -> {
+                feederStopperPower = -Constants.Feeder.FEEDER_HOLD_POWER
+                feederInPower = Constants.Feeder.FEEDER_MAX_POWER
+            }
+            FeederState.HOLD -> {
+                feederStopperPower = -Constants.Feeder.FEEDER_HOLD_POWER
+                feederInPower = Constants.Feeder.FEEDER_HOLD_POWER
+            }
+            FeederState.SHOOT -> {
```

Comment:
Similar thing with the `SHOOT` state -- we don't want to just keep dumping balls into the shooter continuously. Ideally, we can "request" one ball from the feeder at a time, even though likely we will request 5 in quick succession. Allowing "requesting" one ball will basically make it possible to guarantee shooter is back at target speed before feeding the next ball in. 

One potential way to do this would be something like run the feeder up until your ball count goes down by one, or potentially look for a current spike on the stopper wheel, or something, then continue running up until that top beam break has something in it again so you're ready to fire the next ball